xc/xenops_server_xen.ml

```diff
@@ -455,8 +455,13 @@ let device_by_id xc xs vm kind domain_selection id =
 			debug "VM = %s; does not exist in domain list" vm;
 			raise (Does_not_exist("domain", vm))
 		| Some frontend_domid ->
-			let devices = Device_common.list_frontends ~xs frontend_domid in
+			(* Ensure that the frontend belongs to the same vm to counter a misbehaving guest *)
+			if (Device_common.string_uuid_of_domid frontend_domid) <> vm then begin
+				let err = Printf.sprintf "xenstore frontend-backend mismatch for vm = %s" vm in
+				raise (Internal_error err)
+			end;
 
+			let devices = Device_common.list_frontends ~xs frontend_domid in
```

Comment:
I think that the best way to fix the problem is to change Device_common.parse_backend_link() used by list_frontends() so that it also receives the frontend domid as a parameter, and uses this known good frontend domid when creating the backend structure instead of the one parsed from the untrusted "backend" key in the guest's frontend xenstore tree.

For example, this

192 let parse_backend_link x =~$
193 >-------match Xstringext.String.split '/' x with~$
194 >------->-------| [ ""; "local"; "domain"; domid; "backend"; kind; _; devid ] ->$
195 >------->------->-------begin$
196 >------->------->------->-------match parse_int domid, parse_kind kind, parse_int devid with$
197 >------->------->------->------->-------| Some domid, Some kind, Some devid ->$
198 >------->------->------->------->------->-------Some { domid = domid; kind = kind; devid = devid }$

becomes then:

192 let parse_backend_link x frontend_domid =~$
193 >-------match Xstringext.String.split '/' x with~$
194 >------->-------| [ ""; "local"; "domain"; _; "backend"; kind; _; devid ] ->$ (* do not trust domid from this untrusted frontend key *)
195 >------->------->-------begin$
196 >------->------->------->-------match parse_int frontend_domid, parse_kind kind, parse_int devid with$
197 >------->------->------->------->-------| Some frontend_domid, Some kind, Some devid ->$
198 >------->------->------->------->------->-------Some { domid = frontend_domid; kind = kind; devid = devid }$

It looks to me that this should be right in all situations, because I can't see a use case in which we would like to trust the frontend keys to provide the backend domid. The kind and devid are local to the vm, so we can trust these two from the backend link above. The worst that could happen if the vm mishandle these is that it would crash by its own fault.